File: mo-system/mo-system-service/src/main/java/ink/ckx/mo/system/service/impl/SysUserRoleServiceImpl.java
package ink.ckx.mo.system.service.impl;

import cn.hutool.core.collection.CollUtil;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import ink.ckx.mo.system.api.model.entity.SysUserRole;
import ink.ckx.mo.system.mapper.SysUserRoleMapper;
import ink.ckx.mo.system.service.SysUserRoleService;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class SysUserRoleServiceImpl extends ServiceImpl<SysUserRoleMapper, SysUserRole> implements SysUserRoleService {

    /**
     * 保存用户角色
     *
     * @param userId
     * @param roleIds
     * @return
     */
    @Override
    public void saveUserRoles(Long userId, List<Long> roleIds) {
        if (userId == null || CollUtil.isEmpty(roleIds)) {
            return;
        }

        // 用户原角色ID集合
        List<Long> userRoleIds = this.list(new LambdaQueryWrapper<SysUserRole>()
                .eq(SysUserRole::getUserId, userId)).stream().map(SysUserRole::getRoleId).collect(Collectors.toList());

        // 新增用户角色
        List<Long> saveRoleIds;
        if (CollUtil.isEmpty(userRoleIds)) {
            saveRoleIds = roleIds;
        } else {
            saveRoleIds = roleIds.stream().filter(roleId -> !userRoleIds.contains(roleId)).collect(Collectors.toList());
        }

        List<SysUserRole> saveUserRoles = saveRoleIds.stream()
                .map(roleId -> new SysUserRole(null, userId, roleId)).collect(Collectors.toList());
        this.saveBatch(saveUserRoles);

        // 删除用户角色
        if (CollUtil.isNotEmpty(userRoleIds)) {
            List<Long> removeRoleIds = userRoleIds.stream().filter(roleId -> !roleIds.contains(roleId)).collect(Collectors.toList());

            if (CollUtil.isNotEmpty(removeRoleIds)) {
                this.remove(new LambdaQueryWrapper<SysUserRole>().eq(SysUserRole::getUserId, userId).in(SysUserRole::getRoleId, removeRoleIds));
            }
        }
    }
}

File: mo-admin/mo-admin-service/src/main/kotlin/ink/ckx/mo/admin/service/impl/SysUserRoleServiceImpl.kt
package ink.ckx.mo.admin.service.impl

import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import ink.ckx.mo.admin.api.model.entity.SysUserRole
import ink.ckx.mo.admin.mapper.SysUserRoleMapper
import ink.ckx.mo.admin.service.SysUserRoleService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class SysUserRoleServiceImpl : ServiceImpl<SysUserRoleMapper, SysUserRole>(), SysUserRoleService {

    /**
     * 保存用户角色
     *
     * @param userId
     * @param roleIds
     * @return
     */
    @Transactional(rollbackFor = [Exception::class])
    override fun saveUserRoles(userId: Long?, roleIds: List<Long?>) {
        // 用户原角色ID集合
        val userRoleIds = ktQuery()
            .eq(SysUserRole::userId, userId)
            .list()
            .map(SysUserRole::roleId)
            .toList()

        // 新增用户角色
        val saveRoleIds: List<Long?> = if (userRoleIds.isEmpty()) {
            roleIds
        } else {
            roleIds.filter { it !in userRoleIds }.toList()
        }
        val saveUserRoles = saveRoleIds.map { SysUserRole(null, userId, it) }.toList()
        this.saveBatch(saveUserRoles)

        // 删除用户角色
        if (userRoleIds.isNotEmpty()) {
            val removeRoleIds = userRoleIds.filter { it !in roleIds }.toList()
            if (removeRoleIds.isNotEmpty()) {
                ktUpdate()
                    .eq(SysUserRole::userId, userId)
                    .`in`(SysUserRole::roleId, removeRoleIds)
                    .remove()
            }
        }
    }
}